### DISCUSSION TITLE
Data visualization in R

### DISCUSSION THREAD

**Discussion by tecunningham** [jotaen/klog#44] [Show and tell]
Here's some basic code to pull JSON data, parse it, and draw a visualization in R.

<img width="631" alt="image" src="https://user-images.githubusercontent.com/14047958/110418575-acce4080-804c-11eb-91cb-769e73224315.png">

I wanted to share for interest, I will try to slowly improve it.

```r
library(tidyverse)
library(lubridate)
library(hms)
library(scales)
library(ggfittext)

read_json= function() {
  json= system('klog json', intern=TRUE)
  entries_parsed= jsonlite::fromJSON(json, flatten=TRUE )
  entries= entries_parsed$records %>% 
    tidyr::unnest(entries, names_repair='universal') %>%
    tidyr::unnest(tags...12, names_repair='universal') %>%
    mutate( summary= gsub('#[a-zA-Z]+[ ]*', '', summary...11, perl=TRUE) ) # remove hashtags
  return(entries)
}

# a function to reverse scale-order, taken from
# https://stackoverflow.com/questions/43625341/reverse-datetime-posixct-data-axis-in-ggplot
c_trans <- function(a, b, breaks = b$breaks, format = b$format) {
  a <- as.trans(a)
  b <- as.trans(b)
  name <- paste(a$name, b$name, sep = "-")
  trans <- function(x) a$trans(b$trans(x))
  inv <- function(x) b$inverse(a$inverse(x))
  trans_new(name, trans, inverse = inv, breaks = breaks, format=format)
}
rev_date <- c_trans("reverse", "time")

entries= read_json()
ggplot(entries,
        aes(
      ymax=floor_date(as.Date(date)),
      ymin=floor_date(as.Date(date)+1),
      xmin=as_hms(paste0(start,':00')),
      xmax=as_hms(paste0(end,':00')),
      label=summary...11)) +
  geom_rect(aes(fill=tags...12))+
  geom_fit_text(aes(label=summary),reflow=TRUE)+
  scale_y_continuous(trans = rev_date) +
  theme_minimal()
```

**Comment by jotaen** [comment]
It’s really amazing how little code it takes to plot something in R… Have never really used it, so interesting to see how it works for a simple use case. Thanks for sharing!

**Comment by tecunningham** [comment]
I think my end-goal for this would be that you can have your text-editor open with two windows: `klog` file on the left, visualization on the right.

You can get the editor to build the `klog` file, which means calling R to generate an HTML file with embedded images. The HTML file could have various views & visualizations -- broken down by month, by project, by time-of-day.

This setup would be very similar to how people normally work with markdown, latex, etc., with source-file on the left, and output on the right.
